app/src/main/java/com/sdp13epfl2021/projmag/database/OfflineProjectDatabase.kt

```diff
@@ -0,0 +1,208 @@
+package com.sdp13epfl2021.projmag.database
+
+import com.sdp13epfl2021.projmag.model.Failure
+import com.sdp13epfl2021.projmag.model.ImmutableProject
+import com.sdp13epfl2021.projmag.model.Success
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import java.io.*
+import java.util.*
+import kotlin.collections.HashMap
+
+private const val TAG = "OfflineProjectDatabase"
+private const val PROJECT_DATA_FILE: String = "project.data"
+private val ID_PATTERN: Regex = Regex("^[a-zA-Z0-9]*\$")
+
+class OfflineProjectDatabase(private val db: ProjectsDatabase, private val projectsDir: File) : ProjectsDatabase {
+
+    private var projectsFiles: Map<ProjectId, File> = emptyMap()
+
+    init {
+        projectsDir.mkdirs()
+        loadProjects()
+
+        db.addProjectsChangeListener { change ->
+            when(change.type) {
+                ProjectChange.Type.ADDED -> saveProject(change.project)
+                ProjectChange.Type.MODIFIED -> saveProject(change.project)
+                ProjectChange.Type.REMOVED -> deleteProject(change.project.id)
+            }
+        }
+
+        db.getAllProjects({ remoteProjects ->
+            GlobalScope.launch(Dispatchers.IO) {
+                remoteProjects.forEach { p -> saveProject(p) }
+                loadProjects()
+            }
+        }, {})
+    }
+
+    @Synchronized
+    private fun loadProjects() {
+        projectsFiles = projectsDir
+            .listFiles()!!
+            .map { child -> child.name to File(child, PROJECT_DATA_FILE) }
+            .filter { (id, data) -> id.matches(ID_PATTERN) && data.exists() && data.isFile }
+            .toMap()
+    }

```

Comment:
please comment long lambdas with natural language so it's a bit easier to understand (even though in this case it's easy to understand)